cloud/scope/powervs_cluster_test.go

```diff
@@ -3043,3 +3043,2423 @@ func TestCreateVPCSubnet(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		vpcSecurityGroup := infrav1beta2.VPCSecurityGroup{
+			Name: ptr.To("securityGroupNamw"),
+		}
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					VPC: &infrav1beta2.ResourceReference{
+						ID: ptr.To("VPCID"),
+					},
+				},
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					VPCSecurityGroups: append([]infrav1beta2.VPCSecurityGroup{}, vpcSecurityGroup),
+				},
+			},
+		}
+
+		mockVPC.EXPECT().GetSecurityGroupByName(gomock.Any()).Return(nil, errors.New("failed to get SecurityGroup"))
+		sg, ruleIDs, err := clusterScope.validateVPCSecurityGroup(vpcSecurityGroup)
+		g.Expect(ruleIDs).To(BeNil())
+		g.Expect(sg).To(BeNil())
+		g.Expect(err).ToNot(BeNil())
+	})
+
+	t.Run("When SecurityGroup by id doesn't exist and return error", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		vpcSecurityGroup := infrav1beta2.VPCSecurityGroup{
+			ID: ptr.To("securityGroupID"),
+		}
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					VPC: &infrav1beta2.ResourceReference{
+						ID: ptr.To("VPCID"),
+					},
+				},
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					VPCSecurityGroups: append([]infrav1beta2.VPCSecurityGroup{}, vpcSecurityGroup),
+				},
+			},
+		}
+
+		mockVPC.EXPECT().GetSecurityGroup(gomock.Any()).Return(nil, nil, errors.New("failed to get SecurityGroup"))
+		sg, ruleIDs, err := clusterScope.validateVPCSecurityGroup(vpcSecurityGroup)
+		g.Expect(ruleIDs).To(BeNil())
+		g.Expect(sg).To(BeNil())
+		g.Expect(err).ToNot(BeNil())
+	})
+
+	t.Run("When SecurityGroup by name exists but is not attached to VPC", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		vpcSecurityGroup := infrav1beta2.VPCSecurityGroup{
+			Name: ptr.To("securityGroupName"),
+		}
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					VPCSecurityGroups: append([]infrav1beta2.VPCSecurityGroup{}, vpcSecurityGroup),
+				},
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					VPC: &infrav1beta2.ResourceReference{
+						ID: ptr.To("VPCID"),
+					},
+				},
+			},
+		}
+
+		securityGroupDetails := &vpcv1.SecurityGroup{Name: ptr.To("securityGroupName"), ID: ptr.To("sgID")}
+		mockVPC.EXPECT().GetSecurityGroupByName(gomock.Any()).Return(securityGroupDetails, nil)
+		sg, ruleIDs, err := clusterScope.validateVPCSecurityGroup(vpcSecurityGroup)
+		g.Expect(ruleIDs).To(BeEmpty())
+		g.Expect(sg).To(BeEquivalentTo(securityGroupDetails))
+		g.Expect(err).To(BeNil())
+	})
+
+	t.Run("When SecurityGroup by name not exist", func(t *testing.T) {
```

Comment:
These are already validated right, you might have meant a different scenario, please rename accordingly.